### ISSUE TITLE
Handle the app version when state syncing

### ISSUE THREAD

**Issue by evan-forbes** [celestiaorg/celestia-app#3818]
In:
- https://github.com/celestiaorg/celestia-app/pull/3808
- https://github.com/celestiaorg/celestia-app/pull/3811

we hacked around the issues of properly passing the app version to the kv store while mounting it for state sync

https://github.com/celestiaorg/celestia-app/blob/ad250dde754082f68d33a4c4435c01e97125d3ec/app/app.go#L789-L813

To close this issue, we should come up with a long term solution that works for all future app versions.

related to:
- https://github.com/celestiaorg/celestia-app/issues/3807

**Comment by rootulp** [comment]
Notes from meeting:
1. Try to implement this in a way that isn't breaking for v2.x.
2. Consider adding these this field:

```diff
type RequestOfferSnapshot struct {
	Snapshot *Snapshot `protobuf:"bytes,1,opt,name=snapshot,proto3" json:"snapshot,omitempty"`
	AppHash  []byte    `protobuf:"bytes,2,opt,name=app_hash,json=appHash,proto3" json:"app_hash,omitempty"`
+	AppVersion uint64
}
```

3. If a node gets a snapshot that doesn't include AppVersion, fall back to the existing approach where we infer it based on `--v2-upgrade-height`

**Comment by cmwaters** [comment]
> If a node gets a snapshot that doesn't include AppVersion, fall back to the existing approach where we infer it based on --v2-upgrade-height

It can also potentially reject that snapshot and tendermint will search for another one but I think it's probably better that we fallback

I'm not sure if we should add it to the request like you've shown or to the `Snapshot` struct. I had a feeling when I looked at the plumbing that having it in the actual snapshot would be easier but would need to double check.

4. Doesn't have to be coupled with this but it would be nice to have it with the multiapp and v3 as I believe this would clean up the initialisation a lot. (Remember we will still be doing this off the v0.34.x branch so regardless v2 will probably end up having the `AppVersion` in the request regardless of whether we use it or not

**Comment by rootulp** [comment]
Let's de-couple 4 from this issue: https://github.com/celestiaorg/celestia-app/issues/3844

**Comment by rootulp** [comment]
Notes from investigation:

- celestia-core calls list snapshots
- cosmos-sdk implements list snapshots
- We can't co-opt the existing snapshot `format` field to use it for app version. Note if we add a field to the snapshot type then we likely must bump https://github.com/celestiaorg/cosmos-sdk/blob/13d3e17430f313fdf4bd2e5fa9b68ae41768f907/snapshots/types/format.go#L3-L6
- We can't co-opt the existing snapshot `metadata` field because it's populated

**Comment by rootulp** [comment]
After #3871 merges to v2.x, we need to forward-port it to main.
